pyhf/__init__.py

```diff
@@ -204,20 +133,18 @@ def set_poi(self,name):
         assert s.stop-s.start == 1
         self.poi_index = s.start
 
-    def add_modifier(self, name, npars, modifier, suggested_init, suggested_bounds):
-        is_constraint = type(modifier) in [histosys_constraint, normsys_constraint, shapesys_constraint]
+    def add_modifier(self, name, modifier):
+        is_constrained = getattr(modifier, 'is_constrained', False)
```

Comment:
`getattr` seems a bit unpythonic, maybe the modifiers should have `is_constrained()` in which they return their class-level attribute

Reply:
Well, if we don't want abstract classes -- then can't really rely on duck-typing. We can just do `modifier.is_constrained` and this raises an error if it's not set. I guess that's fine.

Reply:
I've added tests for this, so at least for our modifiers, tests will catch this.